enhancements/api-review/add-ClusterImagePolicy-and-ImagePolicy-for-Signature-Verification.md

```diff
@@ -192,6 +196,46 @@ type FulcioCAWithRekor struct {
 	FulcioSubject PolicyFulcioSubject `json:"fulcioSubject,omitempty"`
 }
 
+// PKI defines the root of trust based on Root CA(s) and corresponding intermediate certificates.
+type PKI struct {
+	// CertificateAuthorityRootsData contains a certificate bundle PEM file containing one or more CA roots
+	// in the PEM format.
+	// caRootsData cannot be set if certificateChainData is set.
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateChainData == ''", message="CertificateAuthorityRoots cannot be set if CertificateChain is set"
+	CertificateAuthorityRootsData []byte `json:"caRootsData,omitempty"`
+
+	// CertificateAuthorityIntermediatesData contains a certificate bundle PEM file containing one or more intermediate certificates
+	// in the PEM format.
+	// caIntermediatesData cannot be set if certificateChainData is set.
+	// caIntermediatesData requires CertificateAuthorityRoots to be set.
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateChainData == ''", message="caIntermediatesData cannot be set if certificateChainData is set"
+	// +kubebuilder:validation:XValidation:rule="self == '' || self.CertificateAuthorityRootsData != ''", message="caIntermediatesData requires caRootsData to be set"
+	CertificateAuthorityIntermediatesData []byte `json:"caIntermediatesData,omitempty"`
+
+	// CertificateIdentity must be set if any of the other fields are set.
+	// +optional
+	// +kubebuilder:validation:XValidation:rule="(self.CertificateChainData == '' && self.CertificateAuthorityRootsData == '' && self.caIntermediatesData == '') || self != ''", message="certificateIdentity must be set if any of certificateChainData, caRootsData, or caIntermediatesData are set"
+	// +kubebuilder:validation:XValidation:rule="self.Email != '' || self.Hostname != '' || self.IPAddress != ''", message="At least one of Email, Hostname, or IPAddress must be set in certificateIdentity"
+	CertificateIdentity *CertificateIdentity `json:"certificateIdentity,omitempty"`
+}
+
+// CertificateIdentity defines the identity type associated with the certificate identity
+type CertificateIdentity struct {
```

Comment:
To be explicit, the proposal is that it will be possible to specify no matching requirements at all?

Compare https://issues.redhat.com/browse/RFE-5444?focusedId=24787146&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-24787146 , that’s only reasonable for a strictly single-purpose CA, and then … if we need a CA at all, we probably _have to_ consider revocation; that means the ability to require Rekor, and possible future extensibility to use non-Rekor timestamps as an alternative.

---

I didn’t look into K8s API evolution rules — can we start with making an identity mandatory, and relax that later?

Reply:
maybe we can keep it `optional` and enforce it using kubebuilder validation rules somehow?

Reply:
that way we don't have to change anything in actual API except the rules.

Reply:
// +kubebuilder:validation:XValidation:rule="self.CertificateAuthorityRootsData == '' || self.PKICertificateSubject != nil", message="pkiCertificateSubject must be set if caRootsData is specified"

Reply:
Per a Slack discussion, this is an alpha API and we can update to `alpha2`, so maybe we don’t need to worry about the change between required/optional too much, _for now_.

Reply:
Marked it `required`, we can graduate the API to v1beta2 to make it `optional` if we wish. cc @wking